doc/man/man3/XmChar.3
.TH XMCHAR 3 2026-07-02 "Motif 2.6.0"
.SH NAME
XmChar, XmCodepoint - Data types representing a Unicode character/codepoint

.SH SYNOPSIS
.nf
#include <Xm/XmChar.h>

\fBtypedef unsigned char *\fRXmChar;
\fBtypedef unsigned int\fR XmCodepoint;
.fi

.SH DESCRIPTION
These types are used to represent Unicode characters in various forms for
utility routines. \fBXmChar\fP is a pointer to the first byte of a UTF-8
encoded character. \fBXmCodepoint\fP is an integer type guaranteed to be
at least 32-bits wide.

.SH MACROS
.IP "\fIXM_INVALID_CODEPOINT\fP" 12
A non-character codepoint used as a replacement for invalid codepoints.
.IP "\fIXM_CODEPOINT_MAX\fP" 12
The largest possible Unicode codepoint (\fR0x10ffff\fP).

.SH SEE ALSO
\fBXmCharLen\fP(3), \fBXmCharToCodepoint\fP(3),
\fBXmCodepointToChar\fP(3), \fBXmCodepointNormalize\fP(3),
\fBXmStringNormalize\fP(3)

.SH AUTHOR
Tim Hentenaar <tim@hentenaar.com>

\"vi:set ft=groff:


.attic/ChangeLog
2017-11-10 11:49:10 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1665 (HEAD, origin/master, origin/HEAD, master)

2017-10-30 23:04:15 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1651

2017-10-30 22:22:31 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1650

2017-10-30 21:47:21 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1652

2017-08-31 06:28:27 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1655

2017-08-28 05:33:53 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1625

2017-08-26 06:12:14 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1656

2017-08-23 04:10:55 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1658

2017-08-22 06:05:00 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1659

2017-08-22 03:27:36 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1660

2017-03-28 01:40:33 +0300 Oleksiy Chernyavskyy 

	* updated version and release data

2017-03-28 01:32:30 +0300 Oleksiy Chernyavskyy 

	* removed redundant includes

2017-03-28 01:11:59 +0300 Oleksiy Chernyavskyy 

	* removed bugfix 1565. See details on http://bugs.motifzone.net/show_bug.cgi?id=1565

2016-11-09 07:19:53 +0200 Alex 

	* obsolete malloc.h changed on stdlib.h

2016-06-11 05:03:21 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-11 05:02:55 +0300 Oleksiy Chernyavskyy 

	* updated release 2.3.6 information

2016-06-11 04:57:54 +0300 Oleksiy Chernyavskyy 

	* fixid parallel building (make -j) error

2016-06-11 04:57:20 +0300 Oleksiy Chernyavskyy 

	* removed automatic configure start from autogen.sh

2016-06-11 04:26:48 +0300 Oleksiy Chernyavskyy 

	* minor fix

2016-06-11 04:21:53 +0300 Oleksiy Chernyavskyy 

	* removed extra dist Imakefile from all Makefile.am files

2016-06-11 04:06:37 +0300 Oleksiy Chernyavskyy 

	* Purified copyright headers (Motif is LGPL covered). Removed many Imakefile's and Imake-related stuff

2016-06-10 23:40:43 +0300 Oleksiy Chernyavskyy 

	* updated release notes and version 2.3.5 -> 2.3.6

2016-06-10 23:27:56 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-10 23:15:00 +0300 Oleksiy Chernyavskyy 

	* bug 1654 fix

2016-06-10 21:58:16 +0300 Oleksiy Chernyavskyy 

	* old patches. need review

2016-03-29 03:48:59 +0300 Oleksiy Chernyavskyy 

	* removed obsolete files

2016-03-17 04:23:17 +0200 Oleksiy Chernyavskyy 

	* release 2.3.5 (tag: release_2_3_5)

2016-03-16 23:07:12 +0200 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-03-16 23:06:24 +0200 Oleksiy Chernyavskyy 

	* added mkchangelog script for automated generation of the ChangeLog

2016-03-16 00:46:49 +0200 Oleksiy Chernyavskyy 

	* Reimplemented bugfix 1565

2016-02-26 03:36:51 +0200 Oleksiy Chernyavskyy 

	* Updated RPM creation instructions

2014-08-20 15:45:41 +0300 Oleksiy Chernyavskyy 

	* fixed bug 1636

2013-08-21 11:49:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1612 (Label size computed wrong within a Form).

2013-07-11 18:14:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1618 (Eliminates 29 warnings of some type in modules from "demos" catalog).

2013-07-11 15:20:54 +0300 Mykola Vshyvkov 

	* Fixed bug #1616 (Eliminates 23 warnings "suggest parentheses around assignment used" in modules from "demos" catalog).

2013-07-11 15:01:50 +0300 Mykola Vshyvkov 

	* Fixed bug #1617 (Eliminates 23 warnings "cast to pointer from integer of different size" in modules from "demos" catalog).

2013-07-11 12:35:30 +0300 Mykola Vshyvkov 

	* Fixed bug #1615 (Eliminates 35 warnings "enumeration value XXX not handled" in modules from "demos" catalog).

2013-07-10 06:55:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1614 (Eliminates 36 warnings "variable may be used uninitialized" in modules from "demos" catalog).

2013-07-10 05:05:28 +0300 Mykola Vshyvkov 

	* Fixed bug #1613 (Eliminates 60 warnings "implicit declaration of function" in modules from "demos" catalog).

2013-07-04 14:08:09 +0300 Mykola Vshyvkov 

	* Fixed bug #1611 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls in modules of "clients" catalog).

2013-07-03 21:27:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1610 (Eliminates 15 warnings of some types in modules of "clients" catalog).

2013-07-03 18:52:21 +0300 Mykola Vshyvkov 

	* Fixed bug #1609 (Eliminates 18 warnings  of "incompatible implicit declaration" in modules of "clients" catalog).

2013-07-03 14:58:22 +0300 Mykola Vshyvkov 

	* Fixed bug #1608 (Eliminates 30 warnings  of "array subscript has type 'char'" in modules of "clients" catalog).

2013-07-03 14:35:04 +0300 Mykola Vshyvkov 

	* Fixed bug #1607 (Eliminates 92 warnings  of "variable may be used uninitialized" in modules of "clients" catalog).

2013-07-03 12:48:55 +0300 Mykola Vshyvkov 

	* Fixed bug #1606 (Eliminates 130 warnings  of "expects argument of type" in modules of "clients" catalog).

2013-07-02 13:40:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 20:52:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 19:07:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1604 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls).

2013-07-01 19:02:25 +0300 Mykola Vshyvkov 

	* Fixed bug #1605 (Eliminates 14 warnings of some types in WML tool).

2013-06-26 23:45:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1603 (Eliminates 18 warnings of some types in some modules).

2013-06-26 19:21:20 +0300 Mykola Vshyvkov 

	* Fixed bug #1602 (Eliminates 6 warnings "implicit declaration of function" in some modules).

2013-06-26 18:36:47 +0300 Mykola Vshyvkov 

	* Fixed bug #1601 (Eliminates 6 warnings: "suggest explicit braces to avoid ambiguous" in some modules).

2013-06-26 17:40:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1600 (Eliminates 26 warnings about using uninitialized variables in some modules).

2013-06-26 15:34:53 +0300 Mykola Vshyvkov 

	* Fixed bug #1599 (Eliminates 14 warnings of incompatible type argument from some modules).

2013-06-25 17:37:02 +0300 Mykola Vshyvkov 

	* Updated ChangeLog.

2013-06-25 13:57:31 +0300 Mykola Vshyvkov 

	* Merge branch 'openmotif_2_3' of ssh://git.code.sf.net/p/motif/code into openmotif_2_3


lib/Xm/XmChar.c
/**
 * Motif
 *
 * Copyright (c) 2026 Tim Hentenaar
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
 */

#include <errno.h>

#include "XmChar.h"
#include "XmCharI.h"
#include "unicode.h"

/**
 * Determine the length of a UTF-8 character based on it's initial
 * four bits (0 for contiuation bytes).
 */
static const unsigned int utf8_len[16] = {
	1, 1, 1, 1, 1, 1, 1, 1,
	0, 0, 0, 0, 2, 2, 3, 4
};

/**
 * Get the length (in bytes) of the given XmChar
 */
unsigned int XmCharLen(const XmChar c)
{
	return utf8_len[(*c & 0xf0) >> 4];
}

/**
 * Decode a XmChar to a Unicode codepoint
 */
XmCodepoint XmCharToCodepoint(const XmChar c)
{
	unsigned int i, len;
	XmCodepoint cp = 0;

	if (!c || !*c)
		return 0;

	if ((len = XmCharLen(c)) == 1)
		return c[0];

	switch (len) {
	case 2: return ((c[0] & 0x1f) << 6) | (c[1] & 0x3f);
	case 3:
		cp = ((c[0] & 0x0f) << 12) | ((c[1] & 0x3f) << 6) | (c[2] & 0x3f);
		if (cp >= 0xd800 && cp <= 0xdfff)
			return XM_INVALID_CODEPOINT;
		return cp;
	default:
		/* Ensure we have the mandatory zero bit */
		if ((c[0] & 0xf8) != 0xf0)
			return XM_INVALID_CODEPOINT;
		return ((c[0] & 0x07) << 18) | ((c[1] & 0x1f) << 12) |
		        ((c[2] & 0x3f) << 6) | (c[3] & 0x3f);
	}
}

/**
 * Encode a Unicode codepoint to XmChar.
 *
 * Note: The resulting XmChar must be freed via XtFree().
 */
XmChar XmCodepointToChar(XmCodepoint cp)
{
	XmChar c;
	unsigned int len = 4;

	/* Use a non-character for out-of-range values / UTF-16 surrogates */
	if (cp > XM_CODEPOINT_MAX || (cp >= 0xd800 && cp <= 0xdfff))
		cp = XM_INVALID_CODEPOINT;

	if (cp < 0x80)         len = 1;
	else if (cp < 0x800)   len = 2;
	else if (cp < 0x10000) len = 3;
	c = (XmChar)XtMalloc(len);

	switch (len) {
	case 1:
		c[0] = cp & 0xff;
		break;
	case 2:
		c[0] = 0xc0 | (cp >> 6);
		c[1] = 0x80 | (cp & 0x3f);
		break;
	case 3:
		c[0] = 0xe0 | (cp  >> 12);
		c[1] = 0x80 | ((cp >> 6) & 0x3f);
		c[2] = 0x80 | (cp & 0x3f);
		break;
	default:
		c[0] = 0xf0 | (cp  >> 18);
		c[1] = 0x80 | ((cp >> 12) & 0x3f);
		c[2] = 0x80 | ((cp >> 6)  & 0x3f);
		c[3] = 0x80 | (cp & 0x3f);
	}

	return c;
}

/**
 * Determine whether a word boundary exists between the two
 * given codepoints based on the tr29 default rules.
 */
Boolean XmCodepointIsWordBoundary(XmCodepoint a, XmCodepoint b)
{
	enum codepoint_props cp;
	enum word_break_props wba, wbb;

	if (a > XM_CODEPOINT_MAX || b > XM_CODEPOINT_MAX)
		return True;

	wba = (UCD_PROPS(a).props & WB_MASK)   >> WB_SHIFT;
	wbb = (UCD_PROPS(b).props & WB_MASK)   >> WB_SHIFT;
	cp  = (UCD_PROPS(b).props & PROP_MASK) >> PROP_SHIFT;

	/* WB3c: Do not break within emoji zwj sequences */
	if (wba == WB_ZWJ && cp == CP_ExtendedPictographic)
		return False;

	return wb_table[wba][wbb];
}

/**
 * Returns True if cp is a Hangul Jamo, syllable, or syllabic block
 */
Boolean XmCodepointIsHangul(XmCodepoint cp)
{
	enum line_break_props lb;

	lb = (UCD_PROPS(cp).props & LB_MASK) >> LB_SHIFT;
	return lb == LB_JL || lb == LB_JV || lb == LB_JT ||
	       lb == LB_H2 || lb == LB_H3;
}

/**
 * Compose two codepoints, if able.
 *
 * Returns the composed codepoint, or XM_INVALID_CODEPOINT on error with
 * errno set to indicate the reason (EINVAL: invalid codepoint passed;
 * ENOTSUP: The two codepoints can't combine).
 */
XmCodepoint XmCodepointCompose(XmCodepoint a, XmCodepoint b)
{
	unsigned short i;
	enum line_break_props lba, lbb;

	if (a > XM_CODEPOINT_MAX || b > XM_CODEPOINT_MAX) {
		errno = EINVAL;
		return XM_INVALID_CODEPOINT;
	}

	if (XmCodepointIsHangul(a) && XmCodepointIsHangul(b)) {
		lba = (UCD_PROPS(a).props & LB_MASK) >> LB_SHIFT;
		lbb = (UCD_PROPS(b).props & LB_MASK) >> LB_SHIFT;

		if (lba == LB_JL && lbb == LB_JV) /* L + V */
			return SBASE + (a - LBASE) * NCOUNT + (b - VBASE) * TCOUNT;

		if (lba == LB_H2 && lbb == LB_JT) /* LV + T */
			return a + (b - TBASE);

		errno = ENOTSUP;
		return XM_INVALID_CODEPOINT;
	}

	/* These lists are small, so a linear probe is good enough */
	for (i = 0; i < UCD_PROPS(a).n_comp; i++) {
		if (comp_table[UCD_PROPS(a).comp + i].comb == b)
			return comp_table[UCD_PROPS(a).comp + i].result;
	}

	errno = ENOTSUP;
	return XM_INVALID_CODEPOINT;
}

/**
 * Decompose a codepoint into the given buffer.
 *
 * \param cp Codepoint to decompose
 * \param buf Buffer to write decomposed codepoints to
 * \param len Length of \a buf (in codepoints)
 *
 * Returns the number of codepoints written to the buffer, or 0 if the
 * codepoint could not be decomposed or if an error occured. errno will
 * be set to EINVAL if the parameters given are invalid; ENOSPC if the
 * buffer has insufficient space.
 *
 * If the codepoint doesn't decompose, it gets copied into the buffer,
 * and 1 is returned.
 */
size_t XmCodepointDecompose(XmCodepoint cp, XmCodepoint *buf, size_t len)
{
	size_t out;
	enum line_break_props lb;

	if (!buf || !len || cp > XM_CODEPOINT_MAX) {
		errno = EINVAL;
		return 0;
	}

	if (XmCodepointIsHangul(cp)) {
		lb = (UCD_PROPS(cp).props & LB_MASK) >> LB_SHIFT;
		if (lb != LB_H2 && lb != LB_H3)
			goto nodecomp;

		if (len < 3)
			goto nospc;

		cp    -= SBASE;
		buf[0] = LBASE + cp / NCOUNT;
		buf[1] = VBASE + (cp % NCOUNT) / TCOUNT;
		buf[2] = TBASE + cp % TCOUNT;
		return 2 + (buf[2] != TBASE);
	}

	/* Decompose according to our table */
	if (!(out = UCD_PROPS(cp).n_decomp))
		goto nodecomp;

	if (out <= len) {
		memcpy(buf, decomp_table + UCD_PROPS(cp).decomp, out * sizeof *buf);
		return out;
	}

nospc:
	errno = ENOSPC;
	return 0;

nodecomp:
	*buf = cp;
	return 1;
}

/**
 * Sort combining marks in-place according to the Canonical Ordering
 * Algorithm. Larger values for ccc go further down the sequence, while
 * batches of similar marks should not be reordered.
 */
static void sort_marks(XmCodepoint *buf, size_t len)
{
	size_t i, j;
	XmCodepoint c;
	unsigned int ccc, ccx = 1;

	for (i = 0; i < len; i++) {
		if (!UCD_PROPS(buf[i]).ccc)
			continue;

		for (j = i; j < len - 1; j++) {
			ccc = UCD_PROPS(buf[j]).ccc;
			ccx = UCD_PROPS(buf[j + 1]).ccc;
			if (!ccc || !ccx)
				break;

			if (ccc > ccx) {
				c = buf[j];
				buf[j] = buf[j + 1];
				buf[j + 1] = c;
			}
		}
	}
}

/**
 * Perform a NFC normalization pass over the given set of codepoints,
 * assuming they're NFD-normal and combining marks have been properly
 * reordered.
 */
static XmCodepoint *normalize_nfc(const XmCodepoint *buf, size_t len,
                                  size_t *len_out)
{
	unsigned int qca, qcb;
	size_t i, out_len = 0;
	XmCodepoint *out = NULL, c, d;

	for (i = 0; i < len; i++) {
		/**
		 * Excluded codepoints (NFC_QC=N).
		 */
		qca = (UCD_PROPS(buf[i]).props & QC_MASK) >> QC_SHIFT;
		if (qca == QC_Exclusion)
			continue;

		/**
		 * NFC_QC=M indicates that this is an isolated combining
		 * mark. Copy it out and continue.
		 */
		if (qca == QC_BackwardDecomp || qca == QC_BackwardCombine) {
			out = (XmCodepoint *)XtRealloc((XtPointer)out, (out_len + 1) * sizeof *out);
			out[out_len++] = buf[i];
			continue;
		}

		/**
		 * Here, we have NFC_QC=Y / NFC_QC=M, and can possibly compose.
		 * If the composition fails, emit the current codepoint, and
		 * continue on.
		 */
		c = buf[i];
		while (i < len) {
			d = (++i < len) ? XmCodepointCompose(c, buf[i]) : XM_INVALID_CODEPOINT;
			if (d == XM_INVALID_CODEPOINT) {
				out = (XmCodepoint *)XtRealloc((XtPointer)out, (out_len + 1) * sizeof *out);
				out[out_len++] = c;
				--i;
				break;
			}

			c = d;
		}
	}

	if (len_out) *len_out = out_len;
	return out;
}

/**
 * Perform a NFD normalization pass over the given set of codepoints.
 */
static XmCodepoint *normalize_nfd(const XmCodepoint *buf, size_t len,
                                  size_t *len_out)
{
	unsigned int qc;
	size_t i, x, decomp_len = 3, out_len = 0;
	XmCodepoint *out = NULL, *decomp;

	decomp = (XmCodepoint *)XtMalloc(3 * sizeof *decomp);
	for (i = 0; i < len; i++) {
		qc = (UCD_PROPS(buf[i]).props & QC_MASK) >> QC_SHIFT;

		/**
		 * NFD_QC=Y indicates this codepoint is already D-normal.
		 */
		if (qc == QC_Standard || qc == QC_BackwardCombine) {
			out = (XmCodepoint *)XtRealloc((XtPointer)out, (out_len + 1) * sizeof *out);
			out[out_len++] = buf[i];
			continue;
		}

again:
		/**
		 * NFD_QC=N here implies this codepoint should decompose.
		 */
		x = XmCodepointDecompose(buf[i], decomp, decomp_len);
		if (!x && errno == ENOSPC) {
			if (decomp_len > 8) {
				/**
				 * This shouldn't happen, but this guards against
				 * runaway if we somehow have corrupted data in our
				 * decomposition table, at the expense of the current
				 * codepoint.
				 */
				continue;
			}

			decomp_len += 2;
			decomp = (XmCodepoint *)XtRealloc((XtPointer)decomp, decomp_len * sizeof *decomp);
			goto again;
		}

		/* Add our decomposition to the out buffer */
		out = (XmCodepoint *)XtRealloc((XtPointer)out, (out_len + x) * sizeof *out);
		memcpy(out + out_len, decomp, x * sizeof *decomp);
		out_len += x;
	}

	sort_marks(out, out_len);
	XtFree((XtPointer)decomp);
	if (len_out) *len_out = out_len;
	return out;
}

/**
 * Canonically normalize a series of codepoints.
 *
 * Returns a newly-allocated buffer containing the normalized form of
 * the codepoints from \a buf, with \a len_out being set to the number of
 * codepoints written to the resultant buffer. NULL will be returned if
 * the input parameters are invalid.
 *
 * For NFC, this function will first apply NFD before performing NFC
 * normalization.
 */
XmCodepoint *XmCodepointNormalize(enum XmCodepointNormalForm form,
                                  const XmCodepoint *buf, size_t len,
                                  size_t *len_out)
{
	size_t norm_len;
	XmCodepoint *norm, *out;

	if (len_out) *len_out   = 0;

	if (!buf || !len)
		return NULL;

	norm = normalize_nfd(buf, len, &norm_len);
	if (form == XM_CODEPOINT_NORM_NFD) {
		if (len_out) *len_out = norm_len;
		return norm;
	}

	out = normalize_nfc(norm, norm_len, len_out);
	XtFree((XtPointer)norm);
	return out;
}

